clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -112,12 +112,11 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
                                                + " on memory allocations.");
 
         ByteBuffer buffer = null;
-        this.lock.lock();
```

Comment:
The read/write of ```closed``` must be in lock so this improvement is not acceptable to ```BufferPool```.

Reply:
good catch!